ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs

```diff
@@ -61,82 +61,81 @@ public void Dispose()
-                    value = new KVObjectValue<ulong>(reader.ReadUInt64(), KVValueType.UInt64);
-                    break;
+                var type = ReadNextNodeType();
 
-                case KV1BinaryNodeType.Float32:
-                    var floatValue = BitConverter.ToSingle(reader.ReadBytes(4), 0);
-                    value = new KVObjectValue<float>(floatValue, KVValueType.FloatingPoint);
+                // Break if we hit the end
+                if (type == KV1BinaryNodeType.End)
                     break;
 
-                case KV1BinaryNodeType.Int64:
-                    value = new KVObjectValue<long>(reader.ReadInt64(), KVValueType.Int64);
-                    break;
+                var name = ReadNullTerminatedString();
+                if (type == KV1BinaryNodeType.ChildObject)
+                {
+                    listener.OnObjectStart(name);
+                    ReadObjectCore();
+                    listener.OnObjectEnd();
+                }
+                else
+                {
+                    KVValue value;
+
+                    switch (type)
+                    {
+                        case KV1BinaryNodeType.String:
+                            value = new KVObjectValue<string>(ReadNullTerminatedString(Encoding.UTF8), KVValueType.String);
+                            break;
+
+                        case KV1BinaryNodeType.WideString:
+                            throw new NotSupportedException("Wide String is not supported.");
+
+                        case KV1BinaryNodeType.Int32:
+                        case KV1BinaryNodeType.Color:
+                        case KV1BinaryNodeType.Pointer:
+                            value = new KVObjectValue<int>(reader.ReadInt32(), KVValueType.Int32);
+                            break;
+
+                        case KV1BinaryNodeType.UInt64:
+                            value = new KVObjectValue<ulong>(reader.ReadUInt64(), KVValueType.UInt64);
+                            break;
+
+                        case KV1BinaryNodeType.Float32:
+                            var floatValue = BitConverter.ToSingle(reader.ReadBytes(4), 0);
+                            value = new KVObjectValue<float>(floatValue, KVValueType.FloatingPoint);
+                            break;
+
+                        case KV1BinaryNodeType.Int64:
+                            value = new KVObjectValue<long>(reader.ReadInt64(), KVValueType.Int64);
+                            break;
+
+                        default:
+                            throw new ArgumentOutOfRangeException(nameof(type));
+                    }
 
-                default:
-                    throw new ArgumentOutOfRangeException(nameof(type));
+                    listener.OnKeyValuePair(name, value);
+                }
             }
-
-            listener.OnKeyValuePair(name, value);
         }
 
-        string ReadNullTerminatedString()
+        string ReadNullTerminatedString(Encoding encoding = null)
```

Comment:
This seems misleading to me because it only works with encoding that are terminated by a single NUL byte. I'm not 100% sure this actually covers UTF-8 correctly, but it definitely doesn't work with UTF-16 or UTF-32. Perhaps this function should just read the null terminated byte stream, and the caller should decode.

Reply:
Moved to a function that just reads the null-terminated byte array, and moved the actual decoding to the callers.